### PR TITLE
Mlebabeclap aniso

### PR DESCRIPTION
**Pull request by kkormann**
## Summary

Changes in MLEBABecLap with the goal of removing the assumption dx=dy=dz to make sure anisotropic cells are correctly treated.

## Additional background

The changes concern the way how Feb is computed.

## Checklist

The proposed changes:
- [x] fix a bug or incorrect behavior in AMReX
- [ ] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

**Comment by WeiqunZhang** [comment]
I did some convergence tests using `amrex/Tests/LinearSolvers/CellEB2` and `inputs.rt.2d`.  The results look good.
```
* development branch
  | n_cell | Level 0 max-norm | Level 0 1-norm | Level 1 max-norm | Level 1 1-norm |
  |--------+------------------+----------------+------------------+----------------|
  |     32 |         0.000132 |       7.45e-06 |         4.69e-05 |       4.97e-06 |
  |     64 |         4.16e-05 |       1.17e-06 |         1.55e-05 |       1.40e-06 |
  |    128 |         1.42e-05 |       3.40e-07 |         4.59e-06 |       3.43e-07 |
  |    256 |         4.43e-06 |       8.21e-08 |         1.35e-06 |       8.57e-08 |
  |    512 |         1.32e-06 |       2.19e-08 |         4.62e-07 |       2.24e-08 |
 
* This PR
  | n_cell | Level 0 max-norm | Level 0 1-norm | Level 1 max-norm | Level 1 1-norm |
  |--------+------------------+----------------+------------------+----------------|
  |     32 |         0.000156 |       8.33e-06 |         4.69e-05 |       4.97e-06 |
  |     64 |         4.62e-05 |       1.26e-06 |         1.55e-05 |       1.40e-06 |
  |    128 |         1.51e-05 |       3.49e-07 |         4.59e-06 |       3.44e-07 |
  |    256 |         4.57e-06 |       8.31e-08 |         1.35e-06 |       8.59e-08 |
  |    512 |         1.34e-06 |       2.21e-08 |         4.69e-07 |       2.26e-08 |
```

Will do more tests.

**Comment by drummerdoc** [comment]
How anisotropic are we talking for those tests?  Just curious

**Comment by WeiqunZhang** [comment]
The test I did had uniform size.  I want to make sure the modification works for uniform size.

**Comment by WeiqunZhang** [comment]
Here are the 3d results.
```
* development, 3d
  | n_cell | Level 0 max-norm | Level 0 1-norm | Level 1 max-norm | Level 1 1-norm |
  |--------+------------------+----------------+------------------+----------------|
  |     16 |          0.00122 |       4.74e-05 |         0.000153 |       4.73e-05 |
  |     32 |         0.000910 |       1.77e-05 |         4.68e-05 |       7.42e-06 |
  |     64 |         0.000396 |       3.82e-06 |         1.55e-05 |       2.08e-06 |
  |    128 |         0.000313 |       1.10e-06 |         4.58e-06 |       4.60e-07 |
 
* This PR, 3d
  | n_cell | Level 0 max-norm | Level 0 1-norm | Level 1 max-norm | Level 1 1-norm |
  |--------+------------------+----------------+------------------+----------------|
  |     16 |          0.00122 |       4.74e-05 |         0.000153 |       4.73e-05 |
  |     32 |         0.000910 |       1.77e-05 |         4.68e-05 |       7.42e-06 |
  |     64 |         0.000396 |       3.82e-06 |         1.55e-05 |       2.08e-06 |
  |    128 |         0.000313 |       1.10e-06 |         4.58e-06 |       4.60e-07 |
```

File: Src/LinearSolvers/MLMG/AMReX_MLEBABecLap.cpp [hunk 1]
```diff
@@ -849,9 +849,19 @@ MLEBABecLap::normalize (int amrlev, int mglev, MultiFab& mf) const
...
                  Real dhz = m_b_scalar*dxinvarray[2]*dxinvarray[2];);
+#if (AMREX_SPACEDIM == 1 )
+    const Real dh = m_b_scalar*dxinvarray[0];
+#endif
+#if (AMREX_SPACEDIM == 2 )
+    const Real dh = m_b_scalar*dxinvarray[0]*dxinvarray[1];
+#endif
+#if (AMREX_SPACEDIM == 3 )
+    const Real dh = m_b_scalar*dxinvarray[0]*dxinvarray[1]*dxinvarray[2];
+#endif
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLEBABecLap.cpp]:
```suggestion
    const Real dh = m_b_scalar*AMREX_D_TERM(dxinvarray[0], *dxinvarray[1], *dxinvarray[2]);
```

File: Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_F.cpp [hunk 2]
```diff
@@ -191,10 +189,19 @@ MLEBABecLap::Fsmooth (int amrlev, int mglev, MultiFab& sol, const MultiFab& rhs,
...
                  const Real dhz = m_b_scalar/(h[2]*h[2]));
+#if (AMREX_SPACEDIM == 1 )
+    const Real dh = m_b_scalar/h[0];
+#endif
+#if (AMREX_SPACEDIM == 2 )
+    const Real dh = m_b_scalar/(h[0]*h[1]);
+#endif
+#if (AMREX_SPACEDIM == 3 )
+    const Real dh = m_b_scalar/(h[0]*h[1]*h[2]);
+#endif
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_F.cpp]:
```suggestion
    const Real dh = m_b_scalar/(AMREX_D_TERM(h[0],*h[1],*h[2]));
```

**Review by WeiqunZhang — CHANGES_REQUESTED** [review]
Could you also update `habec_ijmat_eb` in `amrex/Src/Extern/HYPRE/AMReX_Habec_2D_K.H` and `AMReX_Habec_3D_K.H`?  You can use `amrex/Tests/LinearSolvers/CellEB2/` for testing.

File: Src/LinearSolvers/MLMG/AMReX_MLEBABecLap.cpp (hunk 1 above)
Comment (kkormann) [reply Src/LinearSolvers/MLMG/AMReX_MLEBABecLap.cpp]: Done.

File: Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_F.cpp (hunk 2 above)
Comment (kkormann) [reply Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_F.cpp]: Done.

**Comment by kkormann** [comment]
Thanks, @WeiqunZhang , for your suggestions. I have implemented the changes. Also, can you point me to where you compute the center of the boundary edge and the volume fraction? I would guess some update might be necessary there too.

File: Src/Extern/HYPRE/AMReX_Habec_2D_K.H [hunk 3]
```diff
@@ -473,12 +473,15 @@ void habec_ijmat_eb (GpuArray<Real,9>& sten, Array4<Int> const& ncols,
...
+	    Real dapy = (apy(i,j,k) - apy(i,j+1,k))*dx[0];
+            Real anorm = std::sqrt(dapx*dapx
+                                 + dapy*dapy);
 
             Real anorminv = Real(1.0)/anorm;
-            Real anrmx = (apx(i,j,k) - apx(i+1,j,k))*anorminv;
-            Real anrmy = (apy(i,j,k) - apy(i,j+1,k))*anorminv;
+            Real anrmx = dapx*anorminv;
+            Real anrmy = dapy*anorminv;
+	    Real scaling = std::sqrt( (anrmx*dx[0])*(anrmx*dx[0]) + (anrmy*dy[1])*(anrmy*dy[1]) );
```
Comment (WeiqunZhang) [inline Src/Extern/HYPRE/AMReX_Habec_2D_K.H]: `dy` -> `dx`

File: Src/Extern/HYPRE/AMReX_Habec_3D_K.H [hunk 4]
```diff
@@ -952,7 +952,7 @@ void habec_ijmat_eb (GpuArray<Real,27>& sten, Array4<Int> const& ncols,
                                                                                - gxyz};
             Array1D<Real,0,7> feb;
             for (int ii=0; ii<8; ii++) {
-                feb(ii) = -phig1(ii) * (ba(i,j,k) * beb(i,j,k) / dg);
+	      feb(ii) = -phig1(ii) * ( anorm * beb(i,j,k) / (dg*scaling) );
```
Comment (WeiqunZhang) [inline Src/Extern/HYPRE/AMReX_Habec_3D_K.H]: It seems that the change is incomplete.  Additionally, we could also remove `ba` from the function parameter.

**Comment by WeiqunZhang** [comment]
> Also, can you point me to where you compute the center of the boundary edge and the volume fraction? I would guess some update might be necessary there too.

For 3D, it's in `set_eb_data` (AMReX_EB2_3D_C.cpp), which is called by `set_eb_cell`, which in turn is called by `build_cells`.  Yes, ideally we need to update that too.  The algorithm is based on https://crd.lbl.gov/assets/pubs_presos/AMCS/ANAG/SciDAC2008-EBGenerate.pdf.

**Comment by kkormann** [comment]
The tests are now fixed. The `set_eb_data` is only updated for the 2D case that I am currently studying. I think it makes sense to merge the 2D update and leave the 3D update for later. Are there any other changes that I should provide?

File: Src/EB/AMReX_EB2_2D_C.cpp [hunk 5]
```diff
@@ -20,13 +21,17 @@ void set_eb_data (const int i, const int j,
...
-    const Real nx = (axm-axp) * (1.0_rt/apnorm);
-    const Real ny = (aym-ayp) * (1.0_rt/apnorm);
+    const Real axm = apx(i  ,j  ,0)*dx[1];
+    const Real axp = apx(i+1,j  ,0)*dx[1];
+    const Real aym = apy(i  ,j  ,0)*dx[0];
+    const Real ayp = apy(i  ,j+1,0)*dx[0];
+    const Real daxp = (axm-axp);
+    const Real dayp = (aym-ayp);
+    const Real apnorm = std::hypot(daxp,dayp) + 1.e-30_rt;
+    const Real nx = daxp * (1.0/apnorm);
```
Comment (WeiqunZhang) [inline Src/EB/AMReX_EB2_2D_C.cpp]: Could you revert the changes here and many other places that removed `_rt`?

Comment (kkormann) [reply Src/EB/AMReX_EB2_2D_C.cpp]: Oh, yes. The merge introduced some troubles. I went through the changes and fixed.

File: Src/Extern/HYPRE/AMReX_Habec_3D_K.H [hunk 6]
```diff
@@ -914,14 +913,15 @@ void habec_ijmat_eb (GpuArray<Real,27>& sten, Array4<Int> const& ncols,
...
         if (beb) {
-            Real anorm = std::sqrt((apx(i,j,k) - apx(i+1,j,k))*(apx(i,j,k) - apx(i+1,j,k))
-                                 + (apy(i,j,k) - apy(i,j+1,k))*(apy(i,j,k) - apy(i,j+1,k))
-                                 + (apz(i,j,k) - apz(i,j,k+1))*(apz(i,j,k) - apz(i,j,k+1)));
+
+            Real dapx = (apx(i,j,k) - apx(i+1,j,k))*dx[1]*dx[2];
+            Real dapy = (apy(i,j,k) - apy(i,j+1,k))*dx[0]*dx[2];
+            Real dapz = (apz(i,j,k) - apz(i,j,k+1))*dx[1]*dx[2];
 
             Real anorminv = Real(1.0)/anorm;
```
Comment (WeiqunZhang) [inline Src/Extern/HYPRE/AMReX_Habec_3D_K.H]: I don't think `anorm` is defined now.

Comment (WeiqunZhang) [inline Src/Extern/HYPRE/AMReX_Habec_3D_K.H]: I don't think `anorm` is defined now.

Comment (kkormann) [reply Src/Extern/HYPRE/AMReX_Habec_3D_K.H]: anorm is defined now.

File: Src/EB/AMReX_EB2_2D_C.cpp [hunk 7]
```diff
@@ -20,87 +21,96 @@ void set_eb_data (const int i, const int j,
...
     bcent(i,j,0,1) = 0.5_rt*(y_xm+y_xp);
     bnorm(i,j,0,0) = nx;
     bnorm(i,j,0,1) = ny;
 
     if (nxabs < tiny || nyabs > almostone) {
-        vfrac(i,j,0) = 0.5_rt*(axm+axp);
+        vfrac(i,j,0) = 0.5_rt*(axm+axp)*dx[0];
         vcent(i,j,0,0) = 0.0_rt;
-        if (vfrac(i,j,0) > almostone) {
+        if (vfrac(i,j,0)/dx[0] > almostone) {
```
Comment (WeiqunZhang) [inline Src/EB/AMReX_EB2_2D_C.cpp]: I think we should change this test to `vfrac(i,j,0)/(dx[0]*dx[1])`

**Review by WeiqunZhang — CHANGES_REQUESTED** [review]

**Comment by WeiqunZhang** [comment]
You also need to merge the development branch into this to fix a warning error for the intel compiler.

File: Src/EB/AMReX_EB2_2D_C.cpp (hunk 7 above)
Comment (kkormann) [reply Src/EB/AMReX_EB2_2D_C.cpp]: Thanks for spotting this, I fixed it (here and some lines below).

**Comment by kkormann** [comment]
> 

I just merged with development.

File: Src/EB/AMReX_EB2_2D_C.cpp [hunk 8]
```diff
@@ -20,87 +21,96 @@ void set_eb_data (const int i, const int j,
...
-
-        aa = nyabs/nx;
+        Real aa = nxabs/ny*dx[0]/dx[1];
+        const Real dxx = x_ym - x_yp;
+        const Real dx2 = dxx * (x_ym + x_yp);
+        const Real dx3 = dxx * (x_ym*x_ym + x_ym*x_yp + x_yp*x_yp);
+        const Real af1 = 0.5_rt*(axm+axp)*dx[0] + aa*0.5_rt*dx2;
+        vcent(i,j,0,0) = -0.125_rt*daxp*dx[0]*dx[0] + aa*(1._rt/6._rt)*dx3;
+
+        aa = nyabs/nx*dx[1]/dx[1];
```
Comment (WeiqunZhang) [inline Src/EB/AMReX_EB2_2D_C.cpp]:
```suggestion
        aa = nyabs/nx*dx[1]/dx[0];
```

File: Src/EB/AMReX_EB2_2D_C.cpp [hunk 9]
```diff
@@ -20,87 +21,96 @@ void set_eb_data (const int i, const int j,
...
     if (nxabs < tiny || nyabs > almostone) {
-        vfrac(i,j,0) = 0.5_rt*(axm+axp);
+        vfrac(i,j,0) = 0.5_rt*(axm+axp)*dx[0];
         vcent(i,j,0,0) = 0.0_rt;
-        if (vfrac(i,j,0) > almostone) {
+        if (vfrac(i,j,0)/(dx[0]*dx[1]) > almostone) {
             vcent(i,j,0,1) = 0.0_rt;
         } else {
-            vcent(i,j,0,1) = (0.125_rt*(ayp-aym) + ny*0.5_rt*bcent(i,j,0,1)*bcent(i,j,0,1)) / (vfrac(i,j,0) + 1.e-30_rt);
+            vcent(i,j,0,1) = (0.125_rt*dayp*dx[1]*dx[1] + ny*dx[0]*0.5_rt*bcent(i,j,0,1)*bcent(i,j,0,1)) / (vfrac(i,j,0) + 1.e-30_rt);
```
Comment (WeiqunZhang) [inline Src/EB/AMReX_EB2_2D_C.cpp]:
```suggestion
            vcent(i,j,0,1) = (-0.125_rt*dayp*dx[1]*dx[1] + ny*dx[0]*0.5_rt*bcent(i,j,0,1)*bcent(i,j,0,1)) / (vfrac(i,j,0) + 1.e-30_rt);
```

File: Src/EB/AMReX_EB2_2D_C.cpp [hunk 10]
```diff
@@ -20,87 +21,96 @@ void set_eb_data (const int i, const int j,
...
         }
     } else if (nyabs < tiny || nxabs > almostone) {
-        vfrac(i,j,0) = 0.5_rt*(aym+ayp);
-        if (vfrac(i,j,0) > almostone) {
+        vfrac(i,j,0) = 0.5_rt*(aym+ayp)*dx[1];
+        if (vfrac(i,j,0)/(dx[0]*dx[1]) > almostone) {
             vcent(i,j,0,0) = 0.0_rt;
         } else {
-            vcent(i,j,0,0) = (0.125_rt*(axp-axm) + nx*0.5_rt*bcent(i,j,0,0)*bcent(i,j,0,0)) / (vfrac(i,j,0) + 1.e-30_rt);
+            vcent(i,j,0,0) = (0.125_rt*daxp*dx[0]*dx[0] + nx*0.5_rt*bcent(i,j,0,0)*bcent(i,j,0,0)) / (vfrac(i,j,0) + 1.e-30_rt);
```
Comment (WeiqunZhang) [inline Src/EB/AMReX_EB2_2D_C.cpp]:
```suggestion
            vcent(i,j,0,0) = (-0.125_rt*daxp*dx[0]*dx[0] + nx*0.5_rt*bcent(i,j,0,0)*bcent(i,j,0,0)) / (vfrac(i,j,0) + 1.e-30_rt);
```

File: Src/EB/AMReX_EB2_2D_C.cpp [hunk 11]
```diff
@@ -20,87 +21,96 @@ void set_eb_data (const int i, const int j,
...
             vfrac(i,j,0) = 0.0_rt;
             vcent(i,j,0,0) = 0.0_rt;
             vcent(i,j,0,1) = 0.0_rt;
         } else {
             vcent(i,j,0,0) *= (1.0_rt/vfrac(i,j,0));
             vcent(i,j,0,1) *= (1.0_rt/vfrac(i,j,0));
-            vcent(i,j,0,0) = amrex::min(amrex::max(vcent(i,j,0,0),Real(-0.5)),Real(0.5));
-            vcent(i,j,0,1) = amrex::min(amrex::max(vcent(i,j,0,1),Real(-0.5)),Real(0.5));
+            vcent(i,j,0,0) = amrex::min(amrex::max(vcent(i,j,0,0),Real(-0.5)*dx[0]),Real(0.5)*dx[0]);
+            vcent(i,j,0,1) = amrex::min(amrex::max(vcent(i,j,0,1),Real(-0.5)*dx[1]),Real(0.5)*dx[1]);
```
Comment (WeiqunZhang) [inline Src/EB/AMReX_EB2_2D_C.cpp]: Maybe we should move the clamping after vcent is divided by dx to avoid roundoff errors.

File: Src/Extern/HYPRE/AMReX_Habec_3D_K.H [hunk 12]
```diff
@@ -914,14 +913,16 @@ void habec_ijmat_eb (GpuArray<Real,27>& sten, Array4<Int> const& ncols,
         }
 
         if (beb) {
-            Real anorm = std::sqrt((apx(i,j,k) - apx(i+1,j,k))*(apx(i,j,k) - apx(i+1,j,k))
-                                 + (apy(i,j,k) - apy(i,j+1,k))*(apy(i,j,k) - apy(i,j+1,k))
-                                 + (apz(i,j,k) - apz(i,j,k+1))*(apz(i,j,k) - apz(i,j,k+1)));
+
+            Real dapx = (apx(i,j,k) - apx(i+1,j,k))*dx[1]*dx[2];
+            Real dapy = (apy(i,j,k) - apy(i,j+1,k))*dx[0]*dx[2];
+            Real dapz = (apz(i,j,k) - apz(i,j,k+1))*dx[1]*dx[2];
```
Comment (WeiqunZhang) [inline Src/Extern/HYPRE/AMReX_Habec_3D_K.H]:
```suggestion
            Real dapx = (apx(i,j,k) - apx(i+1,j,k));
            Real dapy = (apy(i,j,k) - apy(i,j+1,k));
            Real dapz = (apz(i,j,k) - apz(i,j,k+1));
```
Since 3D has not been implemented, could we just remove those `dx`s?

Comment (WeiqunZhang) [reply Src/Extern/HYPRE/AMReX_Habec_3D_K.H]: Never mind. Maybe we could keep them for consistence because similar changes have been made in MLMG.

File: Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_2D_K.H [hunk 13]
```diff
@@ -360,7 +356,7 @@ void mlebabeclap_ebflux (int i, int j, int k, int n,
             +       (            - gy*sy - gx*gy*sx*sy) * x(i ,jj,k,n)
             +       (                    + gx*gy*sx*sy) * x(ii,jj,k,n) ;
 
-        Real dphidn = dhx*(phib-phig)/dg;
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_2D_K.H:363]: Is this right?  If `dhx` is removed, this will have the wrong unit.

File: Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_3D_K.H [hunk 14]
```diff
@@ -458,7 +461,7 @@ void mlebabeclap_adotx (Box const& box, Array4<Real> const& y,
             y(i,j,k,n) = alpha*a(i,j,k)*x(i,j,k,n) + (1.0/kappa) *
                 (dhx*(apxm*fxm - apxp*fxp) +
                  dhy*(apym*fym - apyp*fyp) +
-                 dhz*(apzm*fzm - apzp*fzp) - dhx*feb);
+                 dhz*(apzm*fzm - apzp*fzp) - dh*feb);
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_3D_K.H]: It could be the case that the changes made here are equivalent to what we had previously when dx == dy == dz.  But it seems to me they are different because `dhx` and `dh` might different units in 3D.  So my suggestion is we revert all the changes in 3D.

File: Src/Extern/HYPRE/AMReX_Habec_3D_K.H (hunk 12 above)
Comment (WeiqunZhang) [reply Src/Extern/HYPRE/AMReX_Habec_3D_K.H]: I just changed my mind again.  I think we should revert all the changes in 3D to be safe.

File: Src/LinearSolvers/MLMG/AMReX_MLEBABecLap.cpp [hunk 15]
```diff
@@ -603,7 +603,7 @@ MLEBABecLap::averageDownCoeffs ()
         auto& fine_b_coeffs = m_b_coeffs[amrlev];
 
         averageDownCoeffsSameAmrLevel(amrlev, fine_a_coeffs, fine_b_coeffs,
-                                      amrex::GetVecOfPtrs(m_eb_b_coeffs[0]));
+                                      amrex::GetVecOfPtrs(m_eb_b_coeffs[amrlev]));
```
Comment (WeiqunZhang) [inline Src/LinearSolvers/MLMG/AMReX_MLEBABecLap.cpp:606]: Thanks for spotting this bug!  Could you make a separate PR fixing this so that we can merge it quickly?

File: Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_2D_K.H (hunk 13 above)
Comment (kkormann) [reply Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_2D_K.H:363]: The scaling is moved inside dg.

File: Src/Extern/HYPRE/AMReX_Habec_3D_K.H (hunk 12 above)
Comment (kkormann) [reply Src/Extern/HYPRE/AMReX_Habec_3D_K.H]: I reverted the changes in 3D (except for the changed interfaces).

File: Src/EB/AMReX_EB2_2D_C.cpp (hunk 11 above)
Comment (kkormann) [reply Src/EB/AMReX_EB2_2D_C.cpp]: Done.

File: Src/EB/AMReX_EB2_2D_C.cpp (hunk 10 above)
Comment (kkormann) [reply Src/EB/AMReX_EB2_2D_C.cpp]: Done.

File: Src/EB/AMReX_EB2_2D_C.cpp (hunk 9 above)
Comment (kkormann) [reply Src/EB/AMReX_EB2_2D_C.cpp]: Done.

File: Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_2D_K.H (hunk 13 above)
Comment (WeiqunZhang) [reply Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_2D_K.H:363]: I am not getting it.  We have
```
Real dapx = ...;
Real dapy = ... same units as dapx;
Real anrmx = dapx / sqrt(dapx*^2 + dapy^2);   //  So this is dimensionless
Real dg = dx_eb / abs(anrmx);   // So dg is dimensionless because dx_eb is dimensionless.  Did I get it right?
```

File: Src/EB/AMReX_EB2_2D_C.cpp [hunk 16]
```diff
@@ -20,72 +21,80 @@ void set_eb_data (const int i, const int j,
...
+            vcent(i,j,0,1) = (-0.125_rt*dayp*dx[1]*dx[1] + ny*dx[0]*0.5_rt*bcent(i,j,0,1)*bcent(i,j,0,1)) / (vfrac(i,j,0)*dx[0]*dx[1] + 1.e-30_rt)  / dx[1];
         }
     } else if (nyabs < tiny || nxabs > almostone) {
-        vfrac(i,j,0) = 0.5_rt*(aym+ayp);
+        vfrac(i,j,0) = 0.5_rt*(aym+ayp)/dx[0];
         if (vfrac(i,j,0) > almostone) {
             vcent(i,j,0,0) = 0.0_rt;
         } else {
-            vcent(i,j,0,0) = (0.125_rt*(axp-axm) + nx*0.5_rt*bcent(i,j,0,0)*bcent(i,j,0,0)) / (vfrac(i,j,0) + 1.e-30_rt);
+            vcent(i,j,0,0) = (-0.125_rt*daxp*dx[0]*dx[0] + nx*0.5_rt*bcent(i,j,0,0)*bcent(i,j,0,0)) / (vfrac(i,j,0)*dx[0]*dx[1] + 1.e-30_rt) / dx[0];
```
Comment (WeiqunZhang) [inline Src/EB/AMReX_EB2_2D_C.cpp]:
```suggestion
            vcent(i,j,0,0) = (-0.125_rt*daxp*dx[0]*dx[0] + nx*dx[1]*0.5_rt*bcent(i,j,0,0)*bcent(i,j,0,0)) / (vfrac(i,j,0)*dx[0]*dx[1] + 1.e-30_rt) / dx[0];
```

**Comment by WeiqunZhang** [comment]
@kkormann Other than the two remaining issues, everything seems fine.

File: Src/EB/AMReX_EB2_2D_C.cpp (hunk 16 above)
Comment (kkormann) [reply Src/EB/AMReX_EB2_2D_C.cpp]: Fixed.

File: Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_2D_K.H (hunk 13 above)
Comment (kkormann) [reply Src/LinearSolvers/MLMG/AMReX_MLEBABecLap_2D_K.H:363]: Sorry I was to quick yesterday. Thinking it through thoroughly I understood the problem. It is now fixed. Thanks!

**Comment by kkormann** [comment]
@WeiqunZhang I think I answered all your request or did I miss something?

**Comment by WeiqunZhang** [comment]
Yes, I think the comments have all been addressed.  But I still like to run some tests. I will try to do it as soon as possible. Thanks.

**Comment by WeiqunZhang** [comment]
Sorry for dragging on this for so long.  This PR probably has the longest conversation.  The good news is the CellEB2 test now shows only roundoff errors, whereas it was much bigger in the beginning.  The bad news is there are still issues in the hypre/petsc setup.  It does not compile.  And, we actually have to restore some of the work you did earlier but was removed later.  The reason for that is the discretization in the external linear solver setup must be consistent with MLMG, otherwise they will not converge.  Unless you want to do it yourself, I can finish this.  I will try to get it done shortly after our next monthly release.  (We usually do not want to merge big changes right before the monthly release.)

**Comment by kkormann** [comment]
> Sorry for dragging on this for so long. This PR probably has the longest conversation. The good news is the CellEB2 test now shows only roundoff errors, whereas it was much bigger in the beginning. The bad news is there are still issues in the hypre/petsc setup. It does not compile. And, we actually have to restore some of the work you did earlier but was removed later. The reason for that is the discretization in the external linear solver setup must be consistent with MLMG, otherwise they will not converge. Unless you want to do it yourself, I can finish this. I will try to get it done shortly after our next monthly release. (We usually do not want to merge big changes right before the monthly release.)

Thanks for taking care of this!

**Review by WeiqunZhang — APPROVED** [review]
